py/dml/codegen.py

```diff
@@ -1624,10 +1733,57 @@ def stmt_return_dml12(stmt, location, scope):
 
 @statement_dispatcher
 def stmt_return(stmt, location, scope):
-    [args] = stmt.args
-    return [codegen_exit(
-        stmt.site, [codegen_expression(arg, location, scope)
-                    for arg in args])]
+    [inits] = stmt.args
+    if isinstance(ExitHandler.current, GotoExit_dml14):
+        outp_typs = [var.ctype() for var in ExitHandler.current.outvars]
+    elif isinstance(ExitHandler.current, ReturnExit):
+        outp_typs = [typ for (_, typ) in ExitHandler.current.outp]
+    else:
+        raise ICE(stmt.site, ("Unexpected ExitHandler "
+                              + f"'{type(ExitHandler.current).__name__}'"))
+
+    if (len(inits) == 1 and inits[0].kind == 'initializer_scalar'
+        and inits[0].args[0].kind == 'apply'):
+        outp = None
+        meth_expr = codegen_expression_maybe_nonvalue(
+            inits[0].args[0].args[0], location, scope)
+        site = meth_expr.site
+        if isinstance(meth_expr, TraitMethodRef):
+            outp = meth_expr.outp
+        elif isinstance(meth_expr, NodeRef):
+            (meth_node, indices) = meth_expr.get_ref()
+            if meth_node.objtype == 'method':
+                outp = meth_node.outp
+        if outp is not None:
+            lscope = Symtab(scope)
+            outarg_syms = [
+                lscope.add_variable(f'tmp{i}', type=typ, site=site,
+                                    static=False, stmt=True)
+                for (i, (_, typ)) in enumerate(outp)]
+            outargs = [mkLocalVariable(site, sym) for sym in outarg_syms]
+            method_invocation = try_codegen_invocation(site, inits, outargs,
+                                                       location, scope)
+            if method_invocation is not None:
+                if len(outargs) != len(outp_typs):
+                    report(ERETARGS(stmt.site, len(outp_typs), len(outargs)))
+                    # avoid control flow errors by falling back to statement
+                    # with no fall-through
+                    return mkAssert(stmt.site,
+                                    mkBoolConstant(stmt.site, False))
+
+
+                return ([sym_declaration(sym) for sym in outarg_syms]
+                        + [method_invocation,
+                           codegen_exit(stmt.site, outargs)])
+
+    if len(inits) != len(outp_typs):
+        report(ERETARGS(stmt.site, len(outp_typs), len(inits)))
+        return [mkAssert(stmt.site, mkBoolConstant(stmt.site, False))]
+
+    return [codegen_exit(stmt.site,
+                         [initast_to_expr(stmt.site, init, typ, location,
```

Comment:
I'm somewhat surprised that this is sufficient -- I had expected that we had to initialize an intermediate local variable. Is this a piece of joy brought by your const fixes?

Reply:
does initast_to_expr mean that it's really easy to permit initializers in method args?